service/src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java

```diff
@@ -276,6 +286,78 @@ public Workspace validateMcWorkspaceAndAction(
     return workspace;
   }
 
+  public void validateWorkspaceState(UUID workspaceUuid) {
+    Workspace workspace = workspaceDao.getWorkspace(workspaceUuid);
+    validateWorkspaceState(workspace);
+  }
+
+  /**
+   * Test that the workspace is in a ready state. This is designed to be used right after
+   * validateWorkspaceAndAction
+   *
+   * @param workspace object describing the workspace, including its current state
+   */
+  public void validateWorkspaceState(Workspace workspace) {
```

Comment:
Is there a reason not to allow READ when workspace is not READY?

Reply:
No. I believe I have applied this only in create, update, and delete cases.
LMK if you see a spot where I got it wrong!

Reply:
grant role - https://github.com/DataBiosphere/terra-workspace-manager/blob/dd-PF-2788-enforce-state/service/src/main/java/bio/terra/workspace/app/controller/WorkspaceApiController.java#L422

clone ws - https://github.com/DataBiosphere/terra-workspace-manager/blob/dd-PF-2788-enforce-state/service/src/main/java/bio/terra/workspace/app/controller/WorkspaceApiController.java#L617

get bucket reference - https://github.com/DataBiosphere/terra-workspace-manager/blob/dd-PF-2788-enforce-state/service/src/main/java/bio/terra/workspace/app/controller/ReferencedGcpResourceController.java#L246

check on source in clone object, bucket, bq table, bq data set & data repo references - ReferencedGcpResourceController.java

- 
in all above we could still make an argument for needed a state check to handle corner cases / concurrent operations. However they are likely too uncommon to need checks.

Reply:
I was aiming to cover create, delete, and update types of operations. I'll check grant/revoke and clone cases and make a PR as needed. Thanks for checking!